### ISSUE TITLE
Anyone having trouble using this via Vite.js ? 

### ISSUE THREAD

**Issue by billyc** [jameslan/libxml2-wasm#59]
Hi, I'm trying to include this in a web project which uses Vite.js to bundle all of the dependencies installed from npm.

Has anyone successfully used this with Vite.js? I am pretty sure that something in my Vite config is the problem, but I cannot find it.

The XmlDocument `.fromString()` and  `.fromBuffer()`static calls both hang forever for me -- even when trying to parse the example text from the readme, `'<note><to>Tove</to></note>'`.

- I have the most recent Vite, 5.4.11
- And it's targeting "es2022" browsers, which includes top-level-wait and pretty much every other recent JS feature.
- I am using 'type: module' but it looks like that earlier issue is already resolved...

**Comment by jameslan** [comment]
Could you provide a simple project that demonstrates this? Hanging is more peculiar than failure.

**Comment by billyc** [comment]
Sure - I created something last night. Project repo is here:
- https://github.com/billyc/libxml2-wasm-testcase
- Clone it and run `npm i` to get a runnable dev site.

Details: 

That repo was created with the latest Vite command line project generator. 

- `npm create vite@latest`
- I chose a Vue project with Typescript. 
 
Vue generated everything you see -- with only necessary modifications to test libxml2-wasm and create a dummy Worker which tries to load the library.

- I added the libxml2-wasm library to package.json and updated the Vite settings to target "esnext" which is required since the library uses await-top-level. 
- Then `npm i` installs all the necessary dependencies

The file `src/LIBXML2WASMEventStreamer.worker.ts` has the offending code which
loads the library and posts a simple message to the app. If you comment out the `XmlDocument.parseString()` call, the worker behaves normally; as is, it hangs and never returns. (For me!)

The file `src/App.vue` contains the brief GUI code; the `mounted()` method has a few 
console.log statements to document spawning the worker and waiting for a result.

I'd be very happy to provide more info if any of this wasn't clear. I'm not sure how to make a simpler bundled Vite project which exhibits the problem; I understand this is a lot to digest if you are not already using Vite or Vue. 

 Thanks so much for even taking a look!

**Comment by jameslan** [comment]
Interestingly, with the invocation of `fromString()` in the code, the line before it(` console.log('Step 3. starting worker');`) won't be executed. That is, the `onmessage` callback won't be called.

**Comment by billyc** [comment]
Yeah ... the only explanation I can think of is that Vite does tree shaking to limit what it actually imports. Removing the call to `fromString()` eliminates the need for Vite to import the library, and thus the worker runs normally.

But retaining the line in the code means... that the import itself is hanging? Because otherwise the "Step 3" log statement should be present.

**Comment by jameslan** [comment]
I’m not familiar with Vue/worker, but if I place a breakpoint after `new MyWoker()` in the `mounted()` event, everything works fine after resuming execution.

Probably it needs to be async/await, or at least delayed execution.

**Comment by jameslan** [comment]
Revisiting this issue. It turns out the issue is with top level await in worker.

Because of the top level await, the worker's setup the `onmessage` event handler is delayed. This means the first message sent to the worker may be as early as before the `onmessage` event handler is ready. In this case, the first message will not be processed by the worker.

So the solution is easy, let the worker sends the first message:

```diff
diff --git a/src/App.vue b/src/App.vue
index 78d842a..a71e92b 100644
--- a/src/App.vue
+++ b/src/App.vue
@@ -21,16 +21,21 @@ const MyComponent = defineComponent({
 
     const worker = new MyWorker() as Worker;
     worker.onmessage = (e: MessageEvent) => {
+      // wait the worker to be ready to send first message
+      if (e.data.initialized) { // first message of initialization
+        worker.postMessage({});
+
+        msg = 'Step 2.  Posted message to worker... should only see this briefly...';
+        this.testMessage = msg;
+        console.log(msg);
+        return;
+      }
+      // other messages
       msg = 'Step 4.  Received message! ' + JSON.stringify(e.data);
       this.testMessage = msg;
       console.log(msg);
     };
 
-    worker.postMessage({});
-
-    msg = 'Step 2.  Posted message to worker... should only see this briefly...';
-    this.testMessage = msg;
-    console.log(msg);
   },
 });
 
diff --git a/src/LIBXML2WASMEventStreamer.worker.ts b/src/LIBXML2WASMEventStreamer.worker.ts
index 682b676..6f7ec38 100644
--- a/src/LIBXML2WASMEventStreamer.worker.ts
+++ b/src/LIBXML2WASMEventStreamer.worker.ts
@@ -10,3 +10,5 @@ onmessage = () => {
 
   postMessage({ success: 1 });
 };
+
+// Initial message: worker is ready
+postMessage({ initialized: true })
```

**Comment by billyc** [comment]
Ahhhh this makes perfect sense! I should be able to test this asap. Thanks so much.
